File: python/darts/darts.py
'''
File: darts.py
This is a general intelligence segmentation framework
File Created: Wednesday, 24th June 2020 3:41:55 pm
Author: enxu (xuen@mokar.com)
-----
Last Modified: Wednesday, 24th June 2020 6:05:52 pm
Modified By: enxu (xuen@mokahr.com)
-----
Copyright 2021 - 2020 Your Company, Moka
'''


class Atom():
    def __init__(self, image, pos):
        """
        image is any thing you want like str\n
        pos is a tuple (start,end) in ori string index \n
        types is any you want make the tags of this atom
        """
        self.image = image
        self.pos = pos
        self.types = set([])

    def __str__(self):
        return str(self.image)


class WCell():
    def __init__(self, atom, lpos, embedding=None, feature=0):
        """
        this is a class use for split\n
        atom is Atom, lpos is (start,end) in atom list \n
        embding is a float or float array \n
        feature is a int use by other tag model
        """
        self.word = atom
        self.pos = lpos
        self.embedding = embedding
        self.feature = feature

    def addTypes(self, types):
        self.word.types |= types

    def getTypes(self):
        return self.word.types

    def __str__(self):
        return "<%s|%s|%s>" % (str(self.word), str(self.pos), str(self.word.types))


class CellMap():
    """
    store all the possobal cell for a string
    """
    class _Cursor():
        def __init__(self, pre, val, next):
            self.pre = pre
            self.val = val
            self.next = next
            self.index = -1

    def __init__(self):
        self.head = CellMap._Cursor(None, WCell(None, (-1, 0)), None)
        self.rownum = 0
        self.colnum = 0
        self.elenum = 0

    def indexMap(self):
        node = self.head
        index = -1
        while node.next is not None:
            node = node.next
            index += 1
            node.index = index

    def iter(self, coursor=None, row=-1):
        coursor = coursor if coursor is not None else self.head
        if row > -1:
            while coursor.next is not None:
                n = coursor.next
                if n.val.pos[0] < row:
                    coursor = n
                    continue
                if n.val.pos[0] != row:
                    break
                yield n
                coursor = n
        else:
            while coursor.next is not None:
                yield coursor.next
                coursor = coursor.next

    def addNext(self, cursor, cell):
        self.rownum = max(self.rownum, cell.pos[0])
        self.colnum = max(self.colnum, cell.pos[1])
        while cursor.next is not None:
            n = cursor.next
            if n.val.pos[0] < cell.pos[0]:
                cursor = n
                continue
            if n.val.pos[0] == cell.pos[0]:
                if n.val.pos[1] < cell.pos[1]:
                    cursor = n
                    continue

                if n.val.pos[1] == cell.pos[1]:
                    n.val.addTypes(cell.getTypes())
                    return n
            m = CellMap._Cursor(cursor, cell, n)
            self.elenum += 1
            cursor.next = m
            n.pre = m
            return m

        cursor.next = CellMap._Cursor(cursor, cell, None)
        self.elenum += 1
        return cursor.next

    def addPre(self, cursor,  cell):
        self.rownum = max(self.rownum, cell.pos[0])
        self.colnum = max(self.colnum, cell.pos[1])
        while cursor.pre != self.head:
            n = cursor.pre
            if n.val.pos[0] > cell.pos[0]:
                cursor = n
                continue
            if n.val.pos[0] == cell.pos[0]:
                if n.val.pos[1] > cell.pos[1]:
                    cursor = n
                    continue
                if n.val.pos[1] == cell.pos[1]:
                    n.val.addTypes(cell.getTypes())
                    return n
            m = CellMap._Cursor(n, cell, cursor)
            self.elenum += 1
            cursor.pre = m
            n.next = m
            return m
        cursor.pre = CellMap._Cursor(self.head, cell, cursor)
        self.elenum += 1
        return cursor.pre

    def addCell(self, cell, coursor=None):
        if coursor is None:
            return self.addNext(self.head, cell)
        if coursor.val.pos[0] < cell.pos[0]:
            return self.addNext(coursor, cell)
        if (coursor.val.pos[0] == cell.pos[0]) and (coursor.val.pos[1] <= cell.pos[1]):
            return self.addNext(coursor, cell)
        return self.addPre(coursor, cell)


class CellPresenter():
    def embed(self, cmap, context):
        """
        set the cmap val data embeding
        cmap:CellMap
        content: list atom
        """
        raise NotImplementedError()


class CellQuantizer(CellPresenter):

    def distance(self, pre, next):
        """
        pre and next is WCell
        return the float distance of from pre to next
        """
        raise NotImplementedError()


class CellRecognizer():

    def read(self, content, cmap):
        """
        recognize the word cell from base string and put it into the map
        content:list atom
        cmap:CellMap
        """
        raise NotImplementedError()


def _split_content(cmap, quantizer, context):
    """
    split the data from content
    content:list atom
    cmap:CellMap
    quantizer:CellQuantizer
    """
    def buildGraph():
        cmap.indexMap()
        quantizer.embed(cmap, context)
        graph, tmp = {}, []
        for coursor in cmap.iter(None, 0):
            tmp.append((coursor.index, 0.0))
        graph[-1] = tmp
        for pre in cmap.iter():
            tmp = []
            for next in cmap.iter(pre, pre.val.pos[1]):
                dist = quantizer.distance(pre.val, next.val)
                if dist > 0.0:
                    tmp.append((next.index, dist))
            if not tmp:
                tmp.append((cmap.elenum, 0.0))
            graph[pre.index] = tmp
        return graph

    def selectPath(graph):
        # init
        sz = len(graph)
        dist, prev = [-1.0]*sz, [-2]*sz
        used = set(range(sz))
        visted = set()
        for node, weight in graph[-1]:
            dist[node] = weight
            visted.add(node)
            prev[node] = -1
        # dijkstra
        while (sz-1) in used:
            mindist, u = float('inf'), 0
            for i in (used & visted):
                if dist[i] < mindist:
                    mindist = dist[i]
                    u = i
            if u == sz-1:
                break
            used.remove(u)
            for node, weight in graph[u]:
                if node not in used:
                    continue
                c = dist[u] + weight
                visted.add(node)
                if dist[node] < 0 or c < dist[node]:
                    dist[node] = c
                    prev[node] = u

        # select
        bestPaths = [sz - 1]
        while bestPaths[-1] > -1:
            bestPaths.append(prev[bestPaths[-1]])
        bestPaths.reverse()
        return bestPaths

    nodes_use = set(selectPath(buildGraph()))
    return iter(cur.val for cur in cmap.iter() if cur.index in nodes_use)


class SentenceSegment():
    def __init__(self, quantizer):
        self.cellRecognizers = []
        assert (quantizer is not None) and isinstance(quantizer, CellQuantizer)
        self.quantizer = quantizer

    def addCellRecognizer(self, recognizer):
        """
        add a cell recognizer
        """
        assert recognizer is not None and isinstance(recognizer, CellRecognizer)
        self.cellRecognizers.append(recognizer)

    def _buildMap(self, atom_list):
        cmap = CellMap()
        _coursor = cmap.head
        for i, atom in enumerate(atom_list):
            _coursor = cmap.addNext(_coursor, WCell(atom, (i, i+1)))
        for recognizer in self.cellRecognizers:
            recognizer.read(atom_list, cmap)
        return cmap

    def smart_cut(self, atom_list, max_mode=False):
        """
        cut the give data in smart mode,\n
        input must Atom List\n
        if max mode is opened this will return all possable word!\n
        return a Wcell List
        """
        if not atom_list:
            return None
        if len(atom_list) == 1:
            return [WCell(atom_list[0], (0, 1))]
        cmap = self._buildMap(atom_list)
        if max_mode:
            return iter(cur.val for cur in cmap.iter())
        return _split_content(cmap, self.quantizer, atom_list)


class Measurement():
    def join(self, vec1, vec2):
        """
        join two vec to one vec\n
        this return a vec and while using for measure method
        """
        raise NotImplementedError()

    def measure(self, vec1, vec2):
        """
        measure the distance of f1 and f2\n
        return distance must >0
        """
        raise NotImplementedError()


class AtomWordVecDic():
    def dimSize(self):
        """
        the dim of this wordvec\n
        return int
        """
        raise NotImplementedError()

    def wordNum(self):
        """
        word num support\n
        return long
        """
        raise NotImplementedError()

    def hasWord(self, atom):
        """
        if has a word in dict\n
        return boolean
        """
        raise NotImplementedError()

    def embeding(self, atom):
        """
        return float array or float,must be not None
        """
        raise NotImplementedError()


class AtomPresenter():
    """
    embeding a atom list
    """

    def embed(self, context):
        """
        content is atom list \n
        return content vec list that use by  next map method
        """
        raise NotImplementedError()

    def map(self, list_embedings, pos):
        """
        pos=(start,end),list_embeding is embed method return \n
        return a vec 
        """
        raise NotImplementedError()


class SmartCellQuantizer(CellQuantizer):
    """
    embeding a cell use wordvec or nural model
    """

    def __init__(self, measurement, atom_wordvec=None, atom_presenter=None):
        assert (measurement is not None) and isinstance(measurement, Measurement)
        assert (atom_wordvec is not None) or (atom_presenter is not None)
        if atom_wordvec is not None:
            assert isinstance(atom_wordvec, AtomWordVecDic)
        if atom_presenter is not None:
            assert isinstance(atom_presenter, AtomPresenter)
        self.measurement = measurement
        self.vecDic = atom_wordvec
        self.presenter = atom_presenter

        def embed(self, cellmap, context):
            # set embeding by model
            if self.presenter is not None:
                list_vec = self.model.embed(context)
                for cur in cellmap.iter():
                    cell = cur.val
                    cell.embedding = self.model.map(list_vec, cell.pos)
            # set embeding by word vec
            if self.vecDic is not None:
                for cur in cellmap.iter():
                    cell = cur.val
                    v = self.vecDic.embeding(cell.word)
                    cell.embedding = self.measurement.join(cell.embedding, v)

        def distance(self, pre,  next):
            return self.measurement.measure(pre.embedding, next.embedding)

File: src/darts/darts.go
package darts

import (
	"fmt"
	"strings"

	"github.com/deckarep/golang-set"
	"gonum.org/v1/gonum/mat"
)

// Atom is  the small cell in split
type Atom struct {
	Image   string     //image of the atom
	St, End int        // [start,end) of this atom in ori string
	Tags    mapset.Set // type is set(string)
}

//String is add a type to atom
func (atom *Atom) String() string {
	return fmt.Sprintf("%s[%d,%d)", atom.Image, atom.St, atom.End)
}

//NewAtom crate the atom
func NewAtom(str *string, start, end int) *Atom {
	atom := new(Atom)
	atom.Image = *str
	atom.St = start
	atom.End = end
	return atom
}

//AddType is add a type to atom
func (atom *Atom) AddType(t string) {
	if atom.Tags == nil {
		atom.Tags = mapset.NewSet()
	}
	atom.Tags.Add(t)
}

//AddTypes is add the types
func (atom *Atom) AddTypes(types mapset.Set) {
	if types == nil {
		return
	}
	if atom.Tags == nil {
		atom.Tags = mapset.NewSet()
	}
	atom.Tags = atom.Tags.Union(types)
}

//AtomList is aarray of atoms
type AtomList struct {
	Str  *string //ori image
	List []*Atom //lsit data
}

//SubAtomList Sub a atonlist
func (alist *AtomList) SubAtomList(start, end int) *Atom {
	var buf strings.Builder
	tags := mapset.NewSet()
	for i := start; i < end; i++ {
		buf.WriteString(alist.List[i].Image)
		if alist.List[i].Tags != nil {
			tags = tags.Intersect(alist.List[i].Tags)
		}
	}
	if buf.Len() > 0 {
		str := buf.String()
		atom := NewAtom(&str, alist.List[start].St, alist.List[end-1].End)
		if tags.Cardinality() > 0 {
			atom.Tags = tags
		}
		return atom
	}
	return nil
}

//NewAtomList make a atom List
func NewAtomList(atoms []*Atom, img *string) *AtomList {
	return &AtomList{Str: img, List: atoms}
}

//StrIterFuc iter the atom list
func (alist *AtomList) StrIterFuc(skipEmpty, skipPos bool) StringIter {
	if !(skipEmpty || skipPos) {
		return func(dfunc func(*string, int) bool) {
			for i, a := range alist.List {
				if dfunc(&a.Image, i) {
					break
				}
			}
		}
	}
	return func(dfunc func(*string, int) bool) {
		se, sp := skipEmpty, skipPos
		for i, a := range alist.List {
			if se && a.Tags != nil && a.Tags.Contains("<EMPTY>") {
				continue
			}
			if sp && a.Tags != nil && a.Tags.Contains("<POS>") {
				continue
			}
			if dfunc(&a.Image, i) {
				break
			}
		}
	}
}

//WCell is the split token parttern
type WCell struct {
	Word   *Atom      //the word in ori string
	St, Et int        //this cell [start,end) in the atom list
	Emb    mat.Vector // matrix os this
	Feat   uint16     //represent which type of this word this is used for tagger
}

//AddTypes add a give types
func (cell *WCell) AddTypes(types mapset.Set) {
	cell.Word.AddTypes(types)
}

//GetTypes give the type to arry
func (cell *WCell) GetTypes() []string {
	if cell.Word.Tags == nil {
		return nil
	}
	tags := cell.Word.Tags.ToSlice()
	stags := make([]string, len(tags))
	for i, t := range tags {
		stags[i] = t.(string)
	}
	return stags
}

//NewWcell create a WCell
func NewWcell(atom *Atom, s, e int) *WCell {
	cell := new(WCell)
	cell.Word = atom
	cell.St = s
	cell.Et = e
	return cell
}

//Cursor map cursor used
type Cursor struct {
	pre, lack *Cursor // pointer of this node pre and lack
	val       *WCell  // val of the node
	idx       int     // index of this course in list
}

func newCur(value *WCell, pre, next *Cursor) *Cursor {
	cur := new(Cursor)
	cur.val = value
	cur.pre = pre
	cur.lack = next
	return cur
}

//CellMap used
type CellMap struct {
	Head               *Cursor //this cellMap size
	rows, colums, size int     // this cellmap countor
}

func newCellMap() *CellMap {
	cmap := new(CellMap)
	cmap.Head = new(Cursor)
	cmap.Head.idx = -1
	cmap.Head.val = new(WCell)
	cmap.Head.val.St = -1
	cmap.Head.val.Et = 0
	return cmap
}

func (cmap *CellMap) indexMap() {
	node := cmap.Head
	node.idx = -1
	index := node.idx
	for node.lack != nil {
		node = node.lack
		index++
		node.idx = index
	}
}

//IterRow is iter the cellmap ,if row<0 tits iter all
func (cmap *CellMap) IterRow(cur *Cursor, row int, dfunc func(*Cursor)) {
	if cur == nil {
		cur = cmap.Head
	}
	if row < 0 { //Iter all if row is negtive
		for cur.lack != nil {
			dfunc(cur.lack)
			cur = cur.lack
		}
		return
	}
	for cur.lack != nil { //Iter the give row from start
		n := cur.lack
		if n.val.St < row {
			cur = n
			continue
		}
		if n.val.St != row {
			break
		}
		dfunc(n)
		cur = n
	}
}

// AddNext do add the cell to give cir next
func (cmap *CellMap) AddNext(cursor *Cursor, cell *WCell) *Cursor {
	if cell.St > cmap.rows {
		cmap.rows = cell.St
	}
	if cell.Et > cmap.colums {
		cmap.colums = cell.Et
	}

	for cursor.lack != nil {
		n := cursor.lack
		if n.val.St < cell.St {
			cursor = n
			continue
		}
		if n.val.St == cell.St {
			if n.val.Et < cell.Et {
				cursor = n
				continue
			}
			if n.val.Et == cell.Et {
				n.val.AddTypes(cell.Word.Tags)
				return n
			}
		}
		m := newCur(cell, cursor, n)
		cmap.size++
		cursor.lack = m
		n.pre = m
		return m
	}
	cursor.lack = newCur(cell, cursor, nil)
	cmap.size++
	return cursor.lack
}

//AddPre add a cell to next
func (cmap *CellMap) AddPre(cur *Cursor, cell *WCell) *Cursor {
	if cell.St > cmap.rows {
		cmap.rows = cell.St
	}
	if cell.Et > cmap.colums {
		cmap.colums = cell.Et
	}
	for cur.pre != cmap.Head {
		n := cur.pre
		if n.val.St > cell.St {
			cur = n
			continue
		}
		if n.val.St == cell.St {
			if n.val.Et > cell.Et {
				cur = n
				continue
			}
			if n.val.Et == cell.Et {
				n.val.AddTypes(cell.Word.Tags)
				return n
			}
		}
		m := newCur(cell, n, cur)
		cmap.size++
		cur.pre = m
		n.lack = m
		return m
	}
	cur.pre = newCur(cell, cmap.Head, cur)
	cmap.size++
	return cur.pre
}

//AddCell add a cell to a cur
func (cmap *CellMap) AddCell(cell *WCell, cur *Cursor) *Cursor {
	if cur == nil {
		return cmap.AddNext(cmap.Head, cell)
	}
	if cur.val.St < cell.St {
		return cmap.AddNext(cur, cell)
	}
	if (cur.val.St == cell.St) && (cur.val.Et <= cell.Et) {
		return cmap.AddNext(cur, cell)
	}
	return cmap.AddPre(cur, cell)
}

//CellRecognizer is code
type CellRecognizer interface {
	//recognizer all Wcell possable in the atomlist
	Read(content *AtomList, cmap *CellMap)
}

//CellQuantizer interface of a prepre
type CellQuantizer interface {
	//set the cmap val data embeding
	Embed(context *AtomList, cmap *CellMap)
	//distance of the pre and next cell
	Distance(pre *WCell, next *WCell) float32
}

func splitContent(cmap *CellMap, quantizer CellQuantizer, context *AtomList) []*WCell {
	buildGraph := func() map[int][][2]float32 {
		cmap.indexMap()
		quantizer.Embed(context, cmap)
		var tmp [][2]float32
		graph := make(map[int][][2]float32)
		cmap.IterRow(nil, 0, func(cur *Cursor) {
			tmp = append(tmp, [2]float32{float32(cur.idx), 0.0})
		})
		graph[-1] = tmp
		cmap.IterRow(nil, -1, func(pre *Cursor) {
			tmp = [][2]float32{}
			cmap.IterRow(pre, pre.val.Et, func(next *Cursor) {
				dist := quantizer.Distance(pre.val, next.val)
				if dist > 0.0 {
					tmp = append(tmp, [2]float32{float32(next.idx), dist})
				}
			})
			if len(tmp) < 1 {
				tmp = append(tmp, [2]float32{float32(cmap.size), 0.0})
			}
			graph[pre.idx] = tmp
		})

		return graph
	}

	selectPath := func(graph map[int][][2]float32) []int {
		// init
		sz := len(graph)
		dist, prev := make([]float32, sz), make([]int, sz)
		dist[0] = -1.0
		for j := 1; j < len(dist); j *= 2 {
			copy(dist[j:], dist[:j])
		}
		prev[0] = -2
		for j := 1; j < len(prev); j *= 2 {
			copy(prev[j:], prev[:j])
		}
		used := mapset.NewSet()
		for j := 1; j < sz; j++ {
			used.Add(j)
		}
		visted := mapset.NewSet()

		for _, nw := range graph[-1] {
			node, weight := int(nw[0]), nw[1]
			dist[node] = weight
			visted.Add(node)
			prev[node] = -1
		}

		// dijkstra
		for used.Contains(sz - 1) {
			minDist, u := float32(3.4e30), 0
			used.Intersect(visted).Each(func(i interface{}) bool {
				idx := i.(int)
				if dist[idx] < minDist {
					minDist = dist[idx]
					u = idx
				}
				return false
			})
			if u == sz-1 {
				break
			}
			used.Remove(u)
			for _, nw := range graph[u] {
				node, weight := int(nw[0]), nw[1]
				if !used.Contains(node) {
					continue
				}
				c := dist[u] + weight
				visted.Add(node)
				if (dist[node] < 0) || (c < dist[node]) {
					dist[node] = c
					prev[node] = u
				}
			}
		}

		// select
		bestPaths := make([]int, 1, sz)
		bestPaths[0] = sz - 1
		for bestPaths[len(bestPaths)-1] > -1 {
			last := bestPaths[len(bestPaths)-1]
			bestPaths = append(bestPaths, prev[last])
		}
		for i, j := 0, len(bestPaths)-1; i < j; i, j = i+1, j-1 {
			bestPaths[i], bestPaths[j] = bestPaths[j], bestPaths[i]
		}
		return bestPaths
	}
	nodeUse := mapset.NewSet()
	for _, idx := range selectPath(buildGraph()) {
		nodeUse.Add(idx)
	}
	result := make([]*WCell, 0, nodeUse.Cardinality())
	cmap.IterRow(nil, -1, func(cur *Cursor) {
		if nodeUse.Contains(cur.idx) {
			result = append(result, cur.val)
		}
	})
	return result
}

//Segment is used for segment
type Segment struct {
	cellRecognizers []CellRecognizer
	quantizer       CellQuantizer
}

//NewSegment make a new segment
func NewSegment(q CellQuantizer) *Segment {
	s := new(Segment)
	s.quantizer = q
	s.cellRecognizers = make([]CellRecognizer, 0, 2)
	return s
}

//AddCellRecognizer add a CellRecognizer
func (seg *Segment) AddCellRecognizer(r CellRecognizer) {
	if r != nil {
		seg.cellRecognizers = append(seg.cellRecognizers, r)
	}
}

func (seg *Segment) buildMap(atomList *AtomList) *CellMap {
	cmap := newCellMap()
	cur := cmap.Head
	for i, atom := range atomList.List {
		cur = cmap.AddNext(cur, NewWcell(atom, i, i+1))
	}

	for _, recognizer := range seg.cellRecognizers {
		recognizer.Read(atomList, cmap)
	}

	return cmap
}

//SmartCut cut a atoml list
func (seg *Segment) SmartCut(atomList *AtomList, maxMode bool) []*WCell {
	if len(atomList.List) < 1 {
		return nil
	}
	if len(atomList.List) == 1 {
		res := make([]*WCell, 1)
		res[0] = NewWcell(atomList.List[0], 0, 1)
		return res
	}
	cmap := seg.buildMap(atomList)
	if maxMode {
		res := make([]*WCell, 0, cmap.size)
		cmap.IterRow(nil, -1, func(cur *Cursor) {
			res = append(res, cur.val)
		})
		return res
	}
	return splitContent(cmap, seg.quantizer, atomList)
}
